internal/dataplane/kong_client_test.go

```diff
@@ -653,6 +671,183 @@ func TestKongClient_ApplyConfigurationEvents(t *testing.T) {
+			},
+		},
+		{
+			name:         "update error with entity errors",
+			updateError:  true,
+			entityErrors: true,
+			expectError:  true,
+			expectEmittingEvents: []string{
+				"Pod: Warning KongConfigurationApplyFailed",
+				"Ingress: Warning KongConfigurationApplyFailed",
+				"Service: Warning KongConfigurationApplyFailed",
+			},
+		},
+		{
+			name:                  "update error with entity errors, fallback configuration applied",
+			fallbackConfiguration: true,
+			updateError:           true,
+			entityErrors:          true,
+			expectError:           true,
+			expectEmittingEvents: []string{
+				"Pod: Warning KongConfigurationApplyFailed",
+				"Ingress: Warning KongConfigurationApplyFailed",
+				"Service: Warning KongConfigurationApplyFailed",
+				"Pod: Normal FallbackKongConfigurationSucceeded",
+			},
+		},
+		{
+			name:                             "update error with entity errors, fallback configuration failures",
+			fallbackConfiguration:            true,
+			fallbackConfigurationUpdateError: true,
+			updateError:                      true,
+			entityErrors:                     true,
+			expectError:                      true,
+			expectEmittingEvents: []string{
+				"Pod: Warning KongConfigurationApplyFailed",
+				"Ingress: Warning KongConfigurationApplyFailed",
+				"Service: Warning KongConfigurationApplyFailed",
+				"Pod: Warning FallbackKongConfigurationApplyFailed",
+				"Ingress: Warning FallbackKongConfigurationApplyFailed",
+			},
+		},
+		{
+			name:                                     "update error with entity errors, fallback translation failures",
+			fallbackConfiguration:                    true,
+			updateError:                              true,
+			entityErrors:                             true,
+			expectError:                              true,
+			fallbackConfigurationTranslationFailures: true,
+			expectEmittingEvents: []string{
+				"Pod: Warning KongConfigurationApplyFailed",
+				"Ingress: Warning KongConfigurationApplyFailed",
+				"Service: Warning KongConfigurationApplyFailed",
+				"Ingress: Warning FallbackKongConfigurationTranslationFailed",
+				"Service: Warning FallbackKongConfigurationTranslationFailed",
+				"Pod: Normal FallbackKongConfigurationSucceeded",
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			updateStrategyResolver := newMockUpdateStrategyResolver(t)
+			configBuilder := newMockKongConfigBuilder()
+			eventRecorder := mocks.NewEventRecorder()
+			lastValidConfigFetcher := &mockKongLastValidConfigFetcher{}
+			kongClient := setupTestKongClient(t, updateStrategyResolver, clientsProvider, configChangeDetector, configBuilder, eventRecorder, lastValidConfigFetcher)
+			kongClient.kongConfig.FallbackConfiguration = tc.fallbackConfiguration
+
+			if tc.translationFailures {
+				configBuilder.translationFailuresToReturn = []failures.ResourceFailure{
+					lo.Must(failures.NewResourceFailure("some reason", testIngress)),
```

Comment:
```suggestion
					require.No(tc, failures.NewResourceFailure("some reason", testIngress)),
```
why not like that? But TBH it doesn't matter here is basically equivalent

Reply:
I'm not aware of any `testify` helper that would be equivalent to `lo.Must` (it panics on err, and returns a value if no error is there). I think it's fine to use such a helper in tests if you're initializing some values with static input.

Reply:
yup, my bad I overlooked returned value. It's good